tsdb/raw.go

```diff
@@ -0,0 +1,950 @@
+				limit:       e.stmt.Limit,
+				offset:      e.stmt.Offset,
+				chunkSize:   e.chunkSize,
+				name:        chunkedOutput.Name,
+				tags:        chunkedOutput.Tags,
+				selectNames: selectFields,
+				aliasNames:  aliasFields,
+				fields:      e.stmt.Fields,
+				c:           out,
+			}
+		}
+		if e.stmt.HasDerivative() {
+			interval, err := derivativeInterval(e.stmt)
+			if err != nil {
+				out <- &models.Row{Err: err}
+				return
+			}
+			rowWriter.transformer = &RawQueryDerivativeProcessor{
+				IsNonNegative:      e.stmt.FunctionCalls()[0].Name == "non_negative_derivative",
+				DerivativeInterval: interval,
+			}
+		}
+
+		// Emit the data via the limiter.
+		if limited := rowWriter.Add(chunkedOutput.Values); limited {
+			// Limit for this tagset was reached, mark it and start draining a new tagset.
+			e.limitTagSet(chunkedOutput.key())
+			continue
+		}
+	}
+
+	close(out)
+}
+
+// mappersDrained returns whether all the executors Mappers have been drained of data.
+func (e *RawExecutor) mappersDrained() bool {
+	for _, m := range e.mappers {
+		if !m.drained {
+			return false
+		}
+	}
+	return true
+}
+
+// nextMapperTagset returns the alphabetically lowest tagset across all Mappers.
+func (e *RawExecutor) nextMapperTagSet() string {
+	tagset := ""
+	for _, m := range e.mappers {
+		if m.bufferedChunk != nil {
+			if tagset == "" {
+				tagset = m.bufferedChunk.key()
+			} else if m.bufferedChunk.key() < tagset {
+				tagset = m.bufferedChunk.key()
+			}
+		}
+	}
+	return tagset
+}
+
+// nextMapperLowestTime returns the lowest minimum time across all Mappers, for the given tagset.
+func (e *RawExecutor) nextMapperLowestTime(tagset string) int64 {
+	minTime := int64(math.MaxInt64)
+	for _, m := range e.mappers {
+		if !m.drained && m.bufferedChunk != nil {
+			if m.bufferedChunk.key() != tagset {
+				continue
+			}
+			t := m.bufferedChunk.Values[len(m.bufferedChunk.Values)-1].Time
+			if t < minTime {
+				minTime = t
+			}
+		}
+	}
+	return minTime
+}
+
+// nextMapperHighestTime returns the highest time across all Mappers, for the given tagset.
+func (e *RawExecutor) nextMapperHighestTime(tagset string) int64 {
+	maxTime := int64(math.MinInt64)
+	for _, m := range e.mappers {
+		if !m.drained && m.bufferedChunk != nil {
+			if m.bufferedChunk.key() != tagset {
+				continue
+			}
+			t := m.bufferedChunk.Values[0].Time
+			if t > maxTime {
+				maxTime = t
+			}
+		}
+	}
+	return maxTime
+}
+
+// tagSetIsLimited returns whether data for the given tagset has been LIMITed.
+func (e *RawExecutor) tagSetIsLimited(tagset string) bool {
+	_, ok := e.limitedTagSets[tagset]
+	return ok
+}
+
+// limitTagSet marks the given taset as LIMITed.
+func (e *RawExecutor) limitTagSet(tagset string) {
+	e.limitedTagSets[tagset] = struct{}{}
+}
+
+// limitedRowWriter accepts raw mapper values, and will emit those values as rows in chunks
+// of the given size. If the chunk size is 0, no chunking will be performed. In addiiton if
```

Comment:
addiiton -> addition (though the typo was probably there when I wrote the comment!)